lib/api/src/lib/stories.ts

```diff
@@ -144,103 +136,89 @@ export const transformStoriesRawToStoriesHash = (
-      if (anyKindMatchesOldHierarchySeparators && !setShowRoots) {
-        warnChangedDefaultHierarchySeparators();
-      }
+    const setShowRoots = typeof showRoots !== 'undefined';
+    if (usesOldHierarchySeparator && !setShowRoots) {
+      warnChangedDefaultHierarchySeparators();
+    }
 
-      let root = '';
-      let groups: string[];
-      const parts: string[] = kind.split('/');
-      // Default showRoots to true if they didn't set it.
-      if ((!setShowRoots || showRoots) && parts.length > 1) {
-        [root, ...groups] = parts;
-      } else {
-        groups = parts;
-      }
+    const groups = kind.split('/').map((part) => part.trim());
+    const root = (!setShowRoots || showRoots) && groups.length > 1 ? [groups.shift()] : [];
+
+    const rootAndGroups = [...root, ...groups].reduce((list, name, index) => {
+      const parent = index > 0 && list[index - 1].id;
+      const id = sanitize(parent ? `${parent}-${name}` : name);
 
-      const rootAndGroups = []
-        .concat(root || [])
-        .concat(groups)
-        .map(toGroup)
-        // Map a bunch of extra fields onto the groups, collecting the path as we go (thus the reduce)
-        .reduce((soFar, group, index, original) => {
-          const { name } = group;
-          const parent = index > 0 && soFar[index - 1].id;
-          const id = sanitize(parent ? `${parent}-${name}` : name);
-          if (parent === id) {
-            throw new Error(
-              dedent`
+      if (parent === id) {
+        throw new Error(
+          dedent`
               Invalid part '${name}', leading to id === parentId ('${id}'), inside kind '${kind}'
 
               Did you create a path that uses the separator char accidentally, such as 'Vue <docs/>' where '/' is a separator char? See https://github.com/storybookjs/storybook/issues/6128
             `
-            );
-          }
+        );
+      }
 
-          if (!!root && index === 0) {
-            const result: Root = {
-              ...group,
-              id,
-              depth: index,
-              children: [],
-              isComponent: false,
-              isLeaf: false,
-              isRoot: true,
-            };
-            return soFar.concat([result]);
-          }
-          const result: Group = {
-            ...group,
-            id,
-            parent,
-            depth: index,
-            children: [],
-            isComponent: false,
-            isLeaf: false,
-            isRoot: false,
-            parameters: {
-              docsOnly: parameters?.docsOnly,
-              viewMode: parameters?.viewMode,
-            },
-          };
-          return soFar.concat([result]);
-        }, [] as GroupsList);
+      if (root.length && index === 0) {
+        list.push({
+          id,
+          name,
+          depth: index,
+          children: [],
+          isComponent: false,
+          isLeaf: false,
+          isRoot: true,
+        });
+      } else {
+        list.push({
```

Comment:
Pushing to the accumulator (`list`) is more efficient than concatting, because that would yield a new array with each iteration.